crates/polars-core/src/fmt.rs

```diff
@@ -964,36 +964,46 @@ fn fmt_struct(f: &mut Formatter<'_>, vals: &[AnyValue]) -> fmt::Result {
 
 impl Series {
     pub fn fmt_list(&self) -> String {
-        match self.len() {
-            0 => "[]".to_string(),
-            1 => format!("[{}]", self.get(0).unwrap()),
-            2 => format!("[{}, {}]", self.get(0).unwrap(), self.get(1).unwrap()),
-            3 => format!(
-                "[{}, {}, {}]",
-                self.get(0).unwrap(),
-                self.get(1).unwrap(),
-                self.get(2).unwrap()
-            ),
-            _ => {
-                let max_items = std::env::var(FMT_TABLE_CELL_LIST_LEN)
-                    .as_deref()
-                    .unwrap_or("")
-                    .parse()
-                    .map_or(3, |n: i64| if n < 0 { self.len() } else { n as usize });
+        if self.is_empty() {
+            return "[]".to_owned();
+        }
+
+        let max_items = std::env::var(FMT_TABLE_CELL_LIST_LEN)
+            .as_deref()
+            .unwrap_or("")
+            .parse()
+            .map_or(3, |n: i64| if n < 0 { self.len() } else { n as usize });
 
+        match max_items {
+            0 => "[…]".to_owned(),
+            _ if max_items >= self.len() => {
                 let mut result = "[".to_owned();
 
-                for (i, item) in self.iter().enumerate() {
+                for i in 0..self.len() {
+                    let item = self.get(i).unwrap();
                     write!(result, "{item}").unwrap();
+                    // this will always leave a trailing ", " after the last item
+                    // but for long lists, this is faster than checking against the length each time
+                    result.push_str(", ");
+                }
+                // remove trailing ", " and replace with closing brace
```

Comment:
if self.is_empty we shouldn't pop right?

Reply:
empty lists shouldn't make it past the guard clause on [line 967](https://github.com/pola-rs/polars/pull/11780/files#diff-c794930ce2f11449d29e5f036c31df358a121a713f0ecbed1cfc07a781f695b0R967), so this loop will always run at least once